File: Structures de donnees complexe/Files/EXERCICE 3/Implementation en JS/Trajet.js
//Definition de la classe Date
class Date
{
	constructor (jour,mois,annee)
	{
		this.jour=jour;
		this.mois=mois;
		this.annee=annee;
	}
}
//Definition de la classe Trajet
class Trajet
{
	constructor (depart,arrivee,date,prix,duree)
	{
		this.depart=depart;
		this.arrivee=arrivee;
		this.date=date;
		this.prix=prix;
		this.duree=duree;
	}
}
//Définition de la classe de noeud
class Node{
	constructor(elt)
	{
		this.data=elt;
		this.next=null;
	}
}
//Définition de la liste de noeuds
class LinkedList{
	constructor()
	{
		this.head=null;
	}
	//Fonction pour ajouter en tête de liste
	save_first(elt)
	{
		var node=new Node(elt);
		node.next=this.head;
		this.head=node;
	}
	//Fonction pour ajouter en fin de liste
	save_end(elt)
	{
		var node=new Node(elt);
		var tmp;
		if (this.head == null){
			this.head=node;
		}
		else
		{
			tmp=this.head;
			while(tmp.next)
			{
				tmp=tmp.next;
			}
			tmp.next=node;	
		}
	}
	//Fonction qui retourner la valeur minimale de la liste
	minimum()
	{
		var tmp,min,elt;
		if (this.head==null)
		{
			console.log("La liste est vide...");
			return null;
			
		}
		else
		{
			tmp=this.head;
			min=tmp.data;
			while(tmp)
			{
				elt=tmp.data;
				if (min.prix > elt.prix)
				{
					min=tmp.data;
				}
				tmp=tmp.next;
			}
			return min;
		}
	}
	//Fonction qui supprime un element de la liste
	remove_elt(elt)
	{
		var prec,cur,i;
		if (this.head==null)
		{
			console.log("La liste est vide...");
		}
		else
		{
			cur=this.head;
			i=0;
			while ((cur.next)&&(cur.data.prix!==elt))
			{
				prec=cur;
				cur=cur.next;
				i++;
			}
			if(cur.data.prix===elt)
			{
				if(i>0)
				{
					prec.next=cur.next;
				}
				else
				{
					this.head=cur.next;
				}
			}	
		}
		
	}
	//Fonction qui retourne la version triée de la liste
	sort_list()
	{
		var new_list=new LinkedList(),elt;
		while(this.head != null)
		{
			elt=this.minimum();
			this.remove_elt(elt.prix);
			new_list.save_first(elt);
		}
		return new_list;
	}
	//Fonction pour afficher la liste
	print_list()
	{
		var tmp=this.head;
		while(tmp)
		{
			console.log(tmp.data.depart+" "+tmp.data.prix);
			tmp=tmp.next;
		}
	}
	//Fonction qui retourne la taille de la liste
	size_of_list()
	{
		var tmp=this.head,nb=0;
		while(tmp)
		{
			nb++;
			tmp=tmp.next;
		}
		return nb
	}
	//Fonction qui indique si un element se trouve dans la liste
	search_elt(elt)
	{
		var tmp,min,elt,trouve;
		trouve=false;
		if (this.head==null)
		{
			console.log("La liste est vide...");
		}
		else
		{
			tmp=this.head;
			while ((tmp.next)&&(tmp.data.depart!==elt))
			{
				tmp=tmp.next;
			}
			if(tmp.data.depart===elt)
			{
				trouve=true;
			}	
		}
		return trouve;
	}
	//Fonction pour lire les données
	read()
	{
		var dat,elt;
		dat =new Date(15,2,2022);
		elt=new Trajet("Beatitude","Ngoa Ekele",dat,400,60);
		this.save_end(elt);
		dat =new Date(13,3,2022);
		elt=new Trajet("Mvan","Ngoa Ekele",dat,300,40);
		this.save_end(elt);
		dat =new Date(15,5,2022);
		elt=new Trajet("Melen","Poste",dat,250,30);
		this.save_end(elt);
		dat =new Date(15,7,2022);
		elt=new Trajet("Education","Ngoa Ekele",dat,100,10);
		this.save_end(elt);
	}
}

//la partir principale
var pile;
pile= new LinkedList();
pile.read();
console.log("La taille de la pile : "+pile.size_of_list());
pile.print_list();
var d="Beatitude";
console.log("Le resultat de la recherche de "+d+" est : "+pile.search_elt(d));
console.log("La taille de la pile : "+pile.size_of_list());
pile=pile.sort_list();
pile.print_list();
console.log("Le minimum est : "+pile.minimum().depart);

File: Structures de donnees complexe/Tableaux/Exercice 3/Implementation en Python/Trajet.py
from random import randrange;
#Classe pour definir la structure de la date
class Date:
	#Fonction d'initialisation
	def __init__(self,jourval=None,moisval=None,anneeval=None):
		self.jour=jourval
		self.mois=moisval
		self.annee=anneeval
#Classe pour definir la structure d'un trajet 
class Trajet:
	#Fonction d'initialisation
	def __init__(self,departval=None,arriveeval=None,date=None,prixval=None,dureeval=None):
		self.depart=departval
		self.arrivee=arriveeval
		self.date=date
		self.prix=prixval
		self.duree=dureeval
class Pile:
	def __init__(self):
		self.data=None
		self.nbelt=0
	def save(self,elt):
		if self.nbelt==0:
			self.data=[elt]
			self.nbelt+=1
		else:
			self.data=[elt]+self.data
			self.nbelt+=1
	#Fonction pour stocker les serie de trajet
	def read(self):
		#1
		date=Date(randrange(1,28),randrange(1,7),2022)
		elt=Trajet("Odza","Mvan",date,300,25)
		self.save(elt)
		#2
		date=Date(randrange(1,28),randrange(1,7),2022)
		elt=Trajet("Beatitude","Ngoa Ekele",date,400,50)
		self.save(elt)
		#3
		date=Date(randrange(1,28),randrange(1,7),2022)
		elt=Trajet("Mvan","Ngoa Ekele",date,350,20)
		self.save(elt)
		#4
		date=Date(randrange(1,28),randrange(1,7),2022)
		elt=Trajet("Ekounou","Mokolo",date,250,50)
		self.save(elt)
		#5
		date=Date(randrange(1,28),randrange(1,7),2022)
		elt=Trajet("Poste","Camer",date,100,5)
		self.save(elt)
		#6
		date=Date(randrange(1,28),randrange(1,7),2022)
		elt=Trajet("Bastos","Education",date,300,50)
		self.save(elt)	
	def print_list(self):
		if self.nbelt==0:
			print("self Vide...")
		else :
			for i in range(self.nbelt):
				elt=self.data[i]
				print("{}\t{}\t{}\t{}".format(elt.depart,elt.arrivee,elt.prix,elt.duree))
	def sort_list(self):
		for i in range(self.nbelt-1):
			mini=i
			for j in range(i+1,self.nbelt):
				if self.data[j].prix<self.data[mini].prix:
					mini=j
			self.data[i],self.data[mini]=self.data[mini],self.data[i]
	def seqSearch(self,elt):
		trouve=False
		i=0
		while(self.data[i].depart!=elt)and(i<self.nbelt):
			i+=1
		if self.data[i].depart==elt:
			trouve=True
		return trouve
pile=Pile()
pile.read()
pile.print_list()
print("\n\n")
pile.sort_list()
pile.print_list()
